otterize/credentials-operator
# Introduce controller that approves CertificateRequests

### Description

When using cert-manager as provider, credentials-operator create `Certificate` objects that should be fulfilled by cert-manager (and generate `Secret` resources accordingly).
cert-manager in turn generates the certificate then creates a `CertificateRequest` resource in order for the issuer to sign that new certificate.
This `CertificateRequest` should be approved before the certificate will actually be signed.
By default, cert-manager's controller automatically approves all requests, but this featured could be turned-off, which is usually the case in enterprise environments.

This PR introduces a new k8s controller that watches for `CertificateRequest` s and approves them if they have the Otterize annotations.


### References

- [CertificateRequest documentation](https://cert-manager.io/v1.1-docs/concepts/certificaterequest/)
- Approver code based on similar controller in [csi-driver-spiffe](https://github.com/cert-manager/csi-driver-spiffe/tree/main/internal/approver/)

### Testing

1. [Disable cert-manager's default approver](https://cert-manager.io/v1.1-docs/concepts/certificaterequest/)
2. Run credentials-operator with the new flag `--cert-manager-approve-requests`
3. See that Certificates & Secrets are successfully created :)


### Checklist

- [x] I have added documentation for new/changed functionality in this PR and in github.com/otterize/docs
